### PR TITLE
counterpart of Python's join function in C++

### PR DESCRIPTION
**Pull request by ferdymercury**
# This Pull request:

## Changes or fixes:

Adds `','.join(strings)` capabilities to `TString.h`.
Right now it is as a variadic argument, but it could be expanded to std::array, std::vector or even TObjArray

## Checklist:

- [x] tested changes locally
- [X] updated the docs (if necessary)

This PR depends on https://github.com/root-project/root/pull/13243

### CODE REVIEW

**Comment by vepadulano** [comment]
Hi @ferdymercury , thank you so much I believe this is a nice addition :smile: ! Most probably it should not stay in `TString.h`, rather in https://github.com/root-project/root/blob/master/core/foundation/inc/ROOT/StringUtils.hxx .

Also, the function uses C++17 features, so the PR needs https://github.com/root-project/root/pull/13243 in first

**Comment by ferdymercury** [comment]
> Hi @ferdymercury , thank you so much I believe this is a nice addition 😄 ! Most probably it should not stay in `TString.h`, rather in https://github.com/root-project/root/blob/master/core/foundation/inc/ROOT/StringUtils.hxx .
> 
> Also, the function uses C++17 features, so the PR needs #13243 in first

Thanks @vepadulano, 13243 is now merged, so I did as  you suggested and moved it to StringUtils

**Comment by ferdymercury** [comment]
one could also use: https://en.cppreference.com/w/cpp/utility/tuple/tuple_cat

**Review by vepadulano — CHANGES_REQUESTED** [review]
Thank you for the addition @ferdymercury ! I believe this is better than `std::tuple_cat` as it doesn't need the input arguments to be `tuples` and it already returns the formatted string. Can I ask you to provide a couple of tests for the new feature?

**Comment by ferdymercury** [comment]
> Thank you for the addition @ferdymercury ! I believe this is better than `std::tuple_cat` as it doesn't need the input arguments to be `tuples` and it already returns the formatted string. Can I ask you to provide a couple of tests for the new feature?

Thanks! I just wrote a couple of tests, but I am not very familiar with the GTest infrastructure, I usually use CTest. So I will need some guidance from you ;)

File: core/foundation/test/testStringUtils.cxx [hunk 1]
```diff
@@ -20,4 +20,23 @@ TEST(StringUtils, Split)
    test(",,,", {}, true);
```
Comment (vepadulano) [inline core/foundation/test/testStringUtils.cxx:20]: You are practically there, instead of putting your new test inside the scope of `TEST(StringUtils, Split)`, you can create a new test function that looks like

```cpp
TEST(StringUtils, Join)
{
    // test body
}
```

File: core/foundation/test/testStringUtils.cxx [hunk 2]
```diff
@@ -20,4 +20,23 @@ TEST(StringUtils, Split)
...
    test(",,a", {"", "", "a"}, false);
    test("", {""}, false);
+
+   // Test that ROOT::Join behaves like str.join from Python.
```
Comment (vepadulano) [inline core/foundation/test/testStringUtils.cxx:24]: This comment goes in the separate test function

File: core/foundation/test/testStringUtils.cxx [hunk 3]
```diff
@@ -20,4 +20,23 @@ TEST(StringUtils, Split)
...
+
+   /*
+   auto test_join_cxx20 = []<typename T>(const std::string res, const std::string& sep, T&&...strings) {
+      auto out = ROOT::Join(sep, strings);
+      EXPECT_EQ(out, ref) << "ROOT::Join(\"" << sep << "\") gave wrong result.";
+   };
+   */
```
Comment (vepadulano) [inline core/foundation/test/testStringUtils.cxx:31]: Remove this part

File: core/foundation/test/testStringUtils.cxx [hunk 4]
```diff
@@ -20,4 +20,23 @@ TEST(StringUtils, Split)
...
+   auto test_join = [](const std::string ref, const std::string& sep, std::vector<std::string> const &strings) {
```
Comment (vepadulano) [inline core/foundation/test/testStringUtils.cxx:32]:
```suggestion
   auto test_join = [](const std::string &ref, const std::string& sep, const std::vector<std::string> &strings) {
```

File: core/foundation/test/testStringUtils.cxx [hunk 5]
```diff
@@ -20,4 +20,23 @@ TEST(StringUtils, Split)
...
+      auto out = ROOT::Join(sep, strings);
+      EXPECT_EQ(out, ref) << "ROOT::Join(\"" << sep << "\") gave wrong result.";
```
Comment (vepadulano) [inline core/foundation/test/testStringUtils.cxx:34]: I don't understand the error message `ROOT::Join(\"" << sep << "\")`, are we sure writing just the `sep` makes the error more clear?

File: core/foundation/test/testStringUtils.cxx [hunk 6]
```diff
@@ -20,4 +20,23 @@ TEST(StringUtils, Split)
...
+   };
+   test_join("apple,orange,banana", ",", {"apple", "orange", "banana"});
+   test_join("apple.orange.banana", ".", {"apple", "orange", "banana"});
+   test_join("appleorangebanana", "", {"apple", "orange", "banana"});
+   test_join("apple", ",", {"apple"});
+   test_join("", ",", {""});
+   test_join("", "", {""});
```
Comment (vepadulano) [inline core/foundation/test/testStringUtils.cxx:41]: Maybe add one test that uses a separator longer than one character

**Review by vepadulano — CHANGES_REQUESTED** [review]
Thank you **so much**! Here are some suggestions to improve the test

File: core/foundation/test/testStringUtils.cxx [hunk 7]
```diff
@@ -20,4 +20,23 @@ TEST(StringUtils, Split)
...
+
+TEST(StringUtils, Join)
+{
+   // Test that ROOT::Join behaves like str.join from Python.
+
+   auto test_join = [](const std::string& ref, const std::string& sep, const std::vector<std::string> &strings) {
+      auto out = ROOT::Join(sep, strings);
+      EXPECT_EQ(out, ref) << "ROOT::Join gave wrong result.";
+   };
+   test_join("apple,orange,banana", ",", {"apple", "orange", "banana"});
```
Comment (vepadulano) [inline core/foundation/test/testStringUtils.cxx:34]: The usage of these tests is different from the function signature. Signature is
```cpp
inline std::string Join(const std::string& sep, T&&...strings) {
```

which means the usage should be `Join(",", "a","b","c")` as I understand it, not `Join(",", {"a", "b", "c"})`

**Review by vepadulano — COMMENTED** [review]
Another fly-by comment that I forgot to add earlier

File: core/foundation/test/testStringUtils.cxx (hunk 7 above)
Comment (vepadulano) [reply core/foundation/test/testStringUtils.cxx:34]: btw, we can totally also add `Join(std::string, std::vector<std::string>)` signature, that can go in a separate PR (not necessarily done by you ofc :) )

Comment (ferdymercury) [reply core/foundation/test/testStringUtils.cxx]: Yep, but then I was confused on how to define the test lambda. I think it needs c++20: https://stackoverflow.com/questions/3575901/can-lambda-functions-be-templated

Comment (ferdymercury) [reply core/foundation/test/testStringUtils.cxx]: Concerning std::vector, I can add it, sure. Maybe via an unpack_vector helper function? https://stackoverflow.com/questions/41980888/how-to-convert-from-stdvector-to-args

Comment (vepadulano) [reply core/foundation/test/testStringUtils.cxx]: One doesn't need to be implemented in terms of the other https://godbolt.org/z/e7fKojY59

File: core/foundation/src/StringUtils.cxx [hunk 8]
```diff
@@ -37,4 +38,21 @@ std::vector<std::string> Split(std::string_view str, std::string_view delims, bo
...
 
+
+/**
+ * @brief Concatenate vector of std::string
+ * @param sep separator inbetween the strings
+ * @param strings vector of strings
+ * @return the sep-delimited concatenation of strings
+ * @see https://stackoverflow.com/questions/9277906/stdvector-to-string-with-custom-delimiter
+ */
+std::string Join(const std::string& sep, std::vector<std::string> strings)
```
Comment (vepadulano) [inline core/foundation/src/StringUtils.cxx:49]:
```suggestion
std::string Join(const std::string& sep, const std::vector<std::string> &strings)
```

File: core/foundation/src/StringUtils.cxx [hunk 9]
```diff
@@ -37,4 +38,21 @@ std::vector<std::string> Split(std::string_view str, std::string_view delims, bo
...
+{
+   if (strings.empty()) return "";
+   return std::accumulate(++std::begin(strings), std::end(strings), strings[0], [&sep](std::string &a, std::string &b)
```
Comment (vepadulano) [inline core/foundation/src/StringUtils.cxx:52]:
```suggestion
   return std::accumulate(std::next(strings), std::end(strings), strings[0], [&sep](const std::string &a, const std::string &b)
```

File: core/foundation/test/testStringUtils.cxx [hunk 10]
```diff
@@ -6,7 +6,7 @@ TEST(StringUtils, Split)
 {
    // Test that ROOT::Split behaves like str.split from Python.
 
-   auto test = [](std::string_view in, std::vector<std::string> const &ref, bool skipEmpty) {
+   auto test = [](std::string_view in, const std::vector<std::string> &ref, bool skipEmpty) {
```
Comment (vepadulano) [inline core/foundation/test/testStringUtils.cxx:9]: Let's not modify other tests that do not belong to this PR

File: core/foundation/test/testStringUtils.cxx [hunk 11]
```diff
@@ -20,4 +20,41 @@ TEST(StringUtils, Split)
...
+   test("apple", ",", {"apple"});
+   test("", ",", {""});
+   test("", "", {""});
+   test("", "", {});
+   test("", ";;", {""});
+
+   /*auto test = []<class...T>(const std::string& ref, const std::string& sep, T&&...strings) {
+      auto out = ROOT::Join(sep, strings);
+      EXPECT_EQ(out, ref) << "ROOT::Join gave wrong result.";
+   };*/
```
Comment (vepadulano) [inline core/foundation/test/testStringUtils.cxx:48]: You can remove these commented lines

File: core/foundation/test/testStringUtils.cxx [hunk 12]
```diff
@@ -20,4 +20,41 @@ TEST(StringUtils, Split)
...
    test(",,a", {"", "", "a"}, false);
    test("", {""}, false);
+
+}
+
+TEST(StringUtils, Join)
+{
+   // Test that ROOT::Join behaves like str.join from Python.
+
+   auto test = [](const std::string& ref, const std::string& sep, const std::vector<std::string> strings) {
```
Comment (vepadulano) [inline core/foundation/test/testStringUtils.cxx:30]:
```suggestion
   auto test = [](const std::string& ref, const std::string& sep, const std::vector<std::string> &strings) {
```

**Review by vepadulano — CHANGES_REQUESTED** [review]

File: core/foundation/inc/ROOT/StringUtils.hxx [hunk 13]
```diff
@@ -23,6 +23,25 @@ namespace ROOT {
 
 std::vector<std::string> Split(std::string_view str, std::string_view delims, bool skipEmpty = false);
 
+std::string Join(const std::string& sep, std::vector<std::string> strings);
+
+/**
+ * @brief Concatenate several std::string
```
Comment (vepadulano) [inline core/foundation/inc/ROOT/StringUtils.hxx:29]: Also one note on the side, ROOT doxygen prefers the `\brief` syntax over the `@brief` syntax (similarly for all other doxygen tags). Could you also modify the docs?

**Review by pcanal — APPROVED** [review]
Thanks.

**Comment by ferdymercury** [comment]
pinging @vepadulano

File: core/foundation/inc/ROOT/StringUtils.hxx [hunk 14]
```diff
@@ -23,6 +23,25 @@ namespace ROOT {
 
 std::vector<std::string> Split(std::string_view str, std::string_view delims, bool skipEmpty = false);
 
+std::string Join(const std::string& sep, const std::vector<std::string>& strings);
```
Comment (guitargeek) [inline core/foundation/inc/ROOT/StringUtils.hxx:26]:
```suggestion
template <class StringCollection_t>
std::string Join(const std::string &sep, StringCollection_t &&strings)
{
   if (strings.empty())
      return "";
   return std::accumulate(std::next(std::begin(strings)), std::end(strings), strings[0],
                          [&sep](auto const &a, auto const &b) { return a + sep + b; });
}
```

Why restrict ourselves to `std::vector<std::string>`? This has several disadvantages:
1. Memory allocation on the heap always necessary because of the `std::vector<>`. We also should deal with `std::span<>`, `std::initialize_list<>`, etc.
2. If you want to concatenate C-style strings, they first have to be copied to a `std::string`

May I suggest the following templated approach?

File: core/foundation/inc/ROOT/StringUtils.hxx [hunk 15]
```diff
@@ -23,6 +23,25 @@ namespace ROOT {
...
+
+/**
+ * \brief Concatenate several std::string
+ * \param[in] sep separator inbetween the strings
+ * \param[in] strings variadic list of strings
+ * \return the sep-delimited concatenation of strings
+ * \see https://stackoverflow.com/a/71983949/7471760
+ */
+template<class...T>
+inline std::string Join(const std::string& sep, T&&...strings) {
```
Comment (guitargeek) [inline core/foundation/inc/ROOT/StringUtils.hxx:36]: What's exactly the usecase of this variadic template method? It's quite restrictive, because the length needs to be known at runtime. Also, the code is hard the read. If you have for example:
```c++
ROOT::Join("foo", "bar", "geek");
```
I find it hard to guess which of the args is the delimiter, the first or the last? I'd have to look up the signature. The `ROOT::Join("foo", {"bar", "geek"})` version would be more intuitive.

Also, I think the variadic template version would be redundant once we have the templated collection version I suggested, because then also then you can also just that one with an initializer list and it will also have no overhead.

File: core/foundation/src/StringUtils.cxx [hunk 16]
```diff
@@ -37,4 +38,21 @@ std::vector<std::string> Split(std::string_view str, std::string_view delims, bo
    return out;
 }
 
+
+/**
+ * \brief Concatenate vector of std::string
+ * \param[in] sep separator inbetween the strings
+ * \param[in] strings vector of strings
+ * \return the sep-delimited concatenation of strings
+ * \see https://stackoverflow.com/questions/9277906/stdvector-to-string-with-custom-delimiter
```
Comment (guitargeek) [inline core/foundation/src/StringUtils.cxx:47]:
```suggestion
```
I don't think it's necessary to quote stackoverflow in the ROOT reference guide :)

File: core/foundation/src/StringUtils.cxx [hunk 17]
```diff
@@ -37,4 +38,21 @@ std::vector<std::string> Split(std::string_view str, std::string_view delims, bo
    return out;
 }
 
+
+/**
+ * \brief Concatenate vector of std::string
+ * \param[in] sep separator inbetween the strings
```
Comment (guitargeek) [inline core/foundation/src/StringUtils.cxx:44]:
```suggestion
 * \param[in] sep Separator inbetween the strings.
```
In doxygen, start your descriptions with an upper-case and end with a period

File: core/foundation/inc/ROOT/StringUtils.hxx (hunk 14 above)
Comment (guitargeek) [reply core/foundation/inc/ROOT/StringUtils.hxx:26]: Also, don't forget to use the `.clang-format` style from the ROOT repo

File: core/foundation/test/testStringUtils.cxx [hunk 18]
```diff
@@ -20,4 +20,37 @@ TEST(StringUtils, Split)
...
+   EXPECT_EQ("apple,orange,banana", ROOT::Join(",", "apple", "orange", "banana"));
+   EXPECT_EQ("apple.orange.banana", ROOT::Join(".", "apple", "orange", "banana"));
+   EXPECT_EQ("apple::orange::banana", ROOT::Join("::", "apple", "orange", "banana"));
+   EXPECT_EQ("appleorangebanana", ROOT::Join("", "apple", "orange", "banana"));
+   EXPECT_EQ("apple,,banana", ROOT::Join(",", "apple", "", "banana"));
+   EXPECT_EQ("apple", ROOT::Join(",", "apple"));
+   EXPECT_EQ("", ROOT::Join(",", ""));
+   EXPECT_EQ("", ROOT::Join("", ""));
+   EXPECT_EQ("", ROOT::Join(""));
+   EXPECT_EQ("", ROOT::Join(";;", ""));
```
Comment (guitargeek) [inline core/foundation/test/testStringUtils.cxx:54]: Probably unnecessary, unless you can come up with a good reason to keep the variadic template version.

**Review by guitargeek — CHANGES_REQUESTED** [review]
Hi @ferdymercury, thanks for your contribution! This is indeed a useful counterpart to `ROOT::Split()` :+1:

I have a few comments though. Mostly about generalizing the method for arbitrary inputs, and a question about the usecase of the variadic template overload.

**Comment by ferdymercury** [comment]
You are welcome, thanks to you for the nice review. I just pushed the branch again with clang-format fixes.
